stratisd_proc_macros/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+            arg_idents.push(ident);
+            (args, arg_idents)
+        },
+    );
+    let stmts = f.block.stmts.drain(..).collect::<Vec<_>>();
+
+    wrapped_sig.inputs = args
+        .into_iter()
+        .map(FnArg::Typed)
+        .collect::<Punctuated<FnArg, Comma>>();
+
+    let method_body_tokens = quote! {
+        #( #stmts )*
+    }
+    .into_iter()
+    .map(process_token_stream)
+    .collect::<TokenStream2>();
+
+    let stmt = parse::<Stmt>(TokenStream::from(quote! {
+        #wrapped_sig {
+            #method_body_tokens
+        }
+    }))
+    .expect("Could not parse generated method as a statement");
+
+    let tokens = quote! { {
+        #stmt
+
+        match #wrapped_ident(#( #arg_idents),*) {
+            Ok(ret) => Ok(ret),
+            Err(e) => {
+                if let StratisError::RollbackError {
+                    ..
+                } = e {
+                    self.action_avail = crate::engine::types::ActionAvailability::ReadOnly;
+                }
+                Err(e)
+            }
+        }
+    } };
+    f.block = parse::<Block>(TokenStream::from(tokens))
+        .expect("Could not parse generated method body as a block");
+}
+
+/// Determine whether a method has the given attribute.
+fn has_attribute(attrs: &mut Vec<Attribute>, attribute: &str) -> bool {
+    let mut return_value = false;
+    let mut index = None;
+    for (i, attr) in attrs.iter().enumerate() {
+        if let Meta::Path(path) = attr
+            .parse_meta()
+            .unwrap_or_else(|_| panic!("Attribute {:?} cannot be parsed", attr))
+        {
+            if path == parse_str(attribute).expect("pool_mutating_action is valid path") {
+                index = Some(i);
+                return_value = true;
+            }
+        }
+    }
+    if let Some(i) = index {
+        attrs.remove(i);
+    }
+    return_value
+}
+
+/// Determine whether a method should be marked as a mutating action based on the attributes.
+///
+/// The attribute that will cause a method to be marked as performing a mutating action is
+/// `#[pool_mutating_action]`.
+fn is_mutating_action(attrs: &mut Vec<Attribute>) -> bool {
+    has_attribute(attrs, "pool_mutating_action")
+}
+
+/// Determine whether a method should be marked as needing to handle failed rollback
+/// based on the attributes.
+///
+/// The attribute that will cause a method to be marked as potentially causing a failed
+/// rollback is `#[pool_rollback]`.
+fn performs_rollback(attrs: &mut Vec<Attribute>) -> bool {
+    has_attribute(attrs, "pool_rollback")
+}
+
+/// Process impl item that was provided to the attribute procedural macro.
+fn process_item(mut item: Item) -> Item {
+    let i = match item {
+        Item::Impl(ref mut i) => i,
+        _ => panic!("This macro can only be applied to impl items"),
+    };
+
+    for impl_item in i.items.iter_mut() {
+        if let ImplItem::Method(ref mut f) = impl_item {
+            if is_mutating_action(&mut f.attrs) {
```

Comment:
There are cases where the ordering does matter, but this is not one of them. This should have the same effect either way.

Reply:
(The main reason for this is that it will be the first statement in the wrapped function as it is now and the added statement can never return a rollback error)